# scrHLAmatrix
For Correcting and Deduplicating scrHLAtag counts and Creating Seurat-compatible Matrices

## Overview
scrHLAmatrix is the accompanying R package for use with data generated using the command line tool [scrHLAtag](https://github.com/furlan-lab/scrHLAtag), which itself is used for processing long-read (PacBio) sequencing data generated from 10x Genomics libraries. In general, scrHLAmatrix uses two kinds of input:
1. The count files of the number of reads that map to the HLA sequence reference and their associated Cell-Barcodes and UMIs. 
2. A related Seurat object where the HLA count matrices generated by scrHLAmatrix will be mapped and embedded. 

## Installation 
In your R console, write the commande:
```
remotes::install_github("https://github.com/furlan-lab/scrHLAmatrix")
```

## Usage
Call the package using:
```
library(scrHLAmatrix)
```
For more information on the various functions and data, as well as examples to run, look into:
```
?HLA_Matrix
?Top_HLA_list_bulk
?Top_HLA_list
?Top_HLA_plot_bulk
?HLA_clusters
?map_HLA_clusters
?HLA_alleles_per_CB
?freq
```
