# scrHLAmatrix
Operating on scrHLAtag output to Determine HLA genotypes of the entity(ies) forming the barcoded cells, and to Correct/Deduplicate HLA genotyping reads, before finally Summarizing them into Seurat-compatible Matrices.

## Overview
scrHLAmatrix is the accompanying R package for use with data generated using the command line tool [scrHLAtag](https://github.com/furlan-lab/scrHLAtag), which itself is used for processing long-read (PacBio) sequencing data generated from 10x Genomics libraries. In general, scrHLAmatrix uses two kinds of input:
1. The count files of the number of reads that map to the HLA sequence reference and their associated Cell-Barcodes and UMIs. 
2. A related Seurat object where the HLA count matrices generated by scrHLAmatrix will be mapped and embedded. 

It is important to note that, first, scrHLAtag must be run on the BAM file(s) from the "corrected" (not the "dedup") step that would have been processed using the single cell [IsoSeq3 pipeline](https://github.com/PacificBiosciences/IsoSeq).

## Installation 
In your R console, write the commande:
```
remotes::install_github("https://github.com/furlan-lab/scrHLAmatrix")
```

## Usage
Call the package using:
```
library(scrHLAmatrix)
```
For more information on the various functions and data, as well as examples to run, look into:
```
?HLA_Matrix
?Top_HLA_list_bulk
?Top_HLA_list
?Top_HLA_plot_bulk
?HLA_clusters
?map_HLA_clusters
?HLA_alleles_per_CB
?freq
```
